File: AocCsharp/Day09/Day09.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using AocCsharp.Utils;

namespace AocCsharp.Day09;

public static class Day09
{
    public static void Run()
    {
        var testInputPath = Path.Combine("..", "inputs", "day09_test.txt");
        var realInputPath = Path.Combine("..", "inputs", "day09.txt");

        TestRunner.RunSolution("Part 1", Part1, testInputPath, realInputPath, expectedTestResult: 50L);
        TestRunner.RunSolution("Part 2", Part2, testInputPath, realInputPath, expectedTestResult: 24L);
    }

    public static long Part1(string input)
    {
        var lines = input.Replace("\r", "").Split('\n', StringSplitOptions.RemoveEmptyEntries);
        if (lines.Length == 0) return 0;

        var redTiles = new List<(int X, int Y)>();
        foreach (var line in lines)
        {
            int commaIndex = line.IndexOf(',');
            if (commaIndex > 0)
            {
                var xSpan = line.AsSpan(0, commaIndex).Trim();
                var ySpan = line.AsSpan(commaIndex + 1).Trim();
                if (int.TryParse(xSpan, out int x) && int.TryParse(ySpan, out int y))
                {
                    redTiles.Add((x, y));
                }
            }
        }

        if (redTiles.Count < 2) return 0;

        long maxArea = 0;

        for (int i = 0; i < redTiles.Count; i++)
        {
            for (int j = i + 1; j < redTiles.Count; j++)
            {
                var tile1 = redTiles[i];
                var tile2 = redTiles[j];

                int width = Math.Abs(tile2.X - tile1.X) + 1;
                int height = Math.Abs(tile2.Y - tile1.Y) + 1;
                long area = (long)width * height;

                maxArea = Math.Max(maxArea, area);
            }
        }

        return maxArea;
    }

    public static long Part2(string input)
    {
        var lines = input.Replace("\r", "").Split('\n', StringSplitOptions.RemoveEmptyEntries);
        if (lines.Length == 0) return 0;

        var redTiles = new List<(int X, int Y)>();
        foreach (var line in lines)
        {
            int commaIndex = line.IndexOf(',');
            if (commaIndex > 0)
            {
                var xSpan = line.AsSpan(0, commaIndex).Trim();
                var ySpan = line.AsSpan(commaIndex + 1).Trim();
                if (int.TryParse(xSpan, out int x) && int.TryParse(ySpan, out int y))
                {
                    redTiles.Add((x, y));
                }
            }
        }

        if (redTiles.Count < 2) return 0;

        // Pre-compute tile set for fast lookup
        var tileSet = new HashSet<(int, int)>(redTiles);
        int n = redTiles.Count;

        long maxArea = 0;

        for (int i = 0; i < n; i++)
        {
            for (int j = i + 1; j < n; j++)
            {
                var tile1 = redTiles[i];
                var tile2 = redTiles[j];

                int rectMinX = Math.Min(tile1.X, tile2.X);
                int rectMaxX = Math.Max(tile1.X, tile2.X);
                int rectMinY = Math.Min(tile1.Y, tile2.Y);
                int rectMaxY = Math.Max(tile1.Y, tile2.Y);

                if (!IsInsideOrOnBoundaryFast((rectMinX, rectMinY), redTiles, tileSet)) continue;
                if (!IsInsideOrOnBoundaryFast((rectMinX, rectMaxY), redTiles, tileSet)) continue;
                if (!IsInsideOrOnBoundaryFast((rectMaxX, rectMinY), redTiles, tileSet)) continue;
                if (!IsInsideOrOnBoundaryFast((rectMaxX, rectMaxY), redTiles, tileSet)) continue;

                bool hasInteriorTile = false;
                for (int k = 0; k < n && !hasInteriorTile; k++)
                {
                    var tile = redTiles[k];
                    if (tile.X > rectMinX && tile.X < rectMaxX && tile.Y > rectMinY && tile.Y < rectMaxY)
                    {
                        hasInteriorTile = true;
                    }
                }

                if (hasInteriorTile) continue;

                bool hasCrossing = false;
                for (int k = 0; k < n && !hasCrossing; k++)
                {
                    var p1 = redTiles[k];
                    var p2 = redTiles[(k + 1) % n];

                    if (SegmentsProperlyIntersect(p1, p2, (rectMinX, rectMinY), (rectMaxX, rectMinY)) ||
                        SegmentsProperlyIntersect(p1, p2, (rectMinX, rectMaxY), (rectMaxX, rectMaxY)) ||
                        SegmentsProperlyIntersect(p1, p2, (rectMinX, rectMinY), (rectMinX, rectMaxY)) ||
                        SegmentsProperlyIntersect(p1, p2, (rectMaxX, rectMinY), (rectMaxX, rectMaxY)))
                    {
                        hasCrossing = true;
                    }
                }

                if (!hasCrossing)
                {
                    long width = rectMaxX - rectMinX + 1;
                    long height = rectMaxY - rectMinY + 1;
                    long area = width * height;
                    maxArea = Math.Max(maxArea, area);
                }
            }
        }

        return maxArea;
    }

    private static bool IsInsideOrOnBoundaryFast((int X, int Y) point, List<(int X, int Y)> polygon, HashSet<(int, int)> tileSet)
    {
        // Fast path: check if point is one of the tiles
        if (tileSet.Contains(point))
        {
            return true;
        }

        // Check if point is on any edge
        int n = polygon.Count;
        for (int i = 0; i < n; i++)
        {
            var p1 = polygon[i];
            var p2 = polygon[(i + 1) % n];

            if (IsPointOnSegment(point, p1, p2))
            {
                return true;
            }
        }

        return IsInsidePolygon(point, polygon);
    }

    private static bool IsInsideOrOnBoundary((int X, int Y) point, List<(int X, int Y)> polygon)
    {
        for (int i = 0; i < polygon.Count; i++)
        {
            var p1 = polygon[i];
            var p2 = polygon[(i + 1) % polygon.Count];

            if (IsPointOnSegment(point, p1, p2))
            {
                return true;
            }
        }

        return IsInsidePolygon(point, polygon);
    }

    private static bool IsPointOnSegment((int X, int Y) point, (int X, int Y) p1, (int X, int Y) p2)
    {
        if (p1.X == p2.X && p1.X == point.X)
        {
            int minY = Math.Min(p1.Y, p2.Y);
            int maxY = Math.Max(p1.Y, p2.Y);
            return point.Y >= minY && point.Y <= maxY;
        }
        if (p1.Y == p2.Y && p1.Y == point.Y)
        {
            int minX = Math.Min(p1.X, p2.X);
            int maxX = Math.Max(p1.X, p2.X);
            return point.X >= minX && point.X <= maxX;
        }
        return false;
    }

    private static bool IsInsidePolygon((int X, int Y) point, List<(int X, int Y)> polygon)
    {
        int intersections = 0;
        int n = polygon.Count;

        for (int i = 0; i < n; i++)
        {
            var p1 = polygon[i];
            var p2 = polygon[(i + 1) % n];

            if ((p1.Y > point.Y) != (p2.Y > point.Y))
            {
                double intersectX = (double)(p2.X - p1.X) * (point.Y - p1.Y) / (p2.Y - p1.Y) + p1.X;
                if (point.X < intersectX)
                {
                    intersections++;
                }
            }
        }

        return (intersections % 2) == 1;
    }

    private static bool SegmentsProperlyIntersect((int X, int Y) p1, (int X, int Y) p2,
        (int X, int Y) p3, (int X, int Y) p4)
    {
        int d1 = Direction(p3, p4, p1);
        int d2 = Direction(p3, p4, p2);
        int d3 = Direction(p1, p2, p3);
        int d4 = Direction(p1, p2, p4);

        if (((d1 > 0 && d2 < 0) || (d1 < 0 && d2 > 0)) &&
            ((d3 > 0 && d4 < 0) || (d3 < 0 && d4 > 0)))
        {
            return true;
        }

        return false;
    }

    private static int Direction((int X, int Y) p1, (int X, int Y) p2, (int X, int Y) p3)
    {
        long val = (long)(p3.Y - p1.Y) * (p2.X - p1.X) - (long)(p2.Y - p1.Y) * (p3.X - p1.X);
        if (val == 0) return 0;
        return (val > 0) ? 1 : -1;
    }
}

File: AocC/day09.c
#include "day09.h"
#include "runner.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

void day09_run(void) {
    const char* test_path = "../inputs/day09_test.txt";
    const char* real_path = "../inputs/day09.txt";

    run_solution("Part 1", day09_part1, test_path, real_path, 50);
    run_solution("Part 2", day09_part2, test_path, real_path, 24);
}

typedef struct {
    int x, y;
} Point;

static int direction(Point p1, Point p2, Point p3) {
    int64_t val = (int64_t)(p3.y - p1.y) * (p2.x - p1.x) -
                  (int64_t)(p2.y - p1.y) * (p3.x - p1.x);
    if (val == 0) return 0;
    return (val > 0) ? 1 : -1;
}

static bool segments_properly_intersect(Point p1, Point p2, Point p3, Point p4) {
    int d1 = direction(p3, p4, p1);
    int d2 = direction(p3, p4, p2);
    int d3 = direction(p1, p2, p3);
    int d4 = direction(p1, p2, p4);

    return ((d1 > 0 && d2 < 0) || (d1 < 0 && d2 > 0)) &&
           ((d3 > 0 && d4 < 0) || (d3 < 0 && d4 > 0));
}

static bool is_point_on_segment(Point p, Point s1, Point s2) {
    if (s1.x == s2.x && s1.x == p.x) {
        int min_y = (s1.y < s2.y) ? s1.y : s2.y;
        int max_y = (s1.y > s2.y) ? s1.y : s2.y;
        return p.y >= min_y && p.y <= max_y;
    }
    if (s1.y == s2.y && s1.y == p.y) {
        int min_x = (s1.x < s2.x) ? s1.x : s2.x;
        int max_x = (s1.x > s2.x) ? s1.x : s2.x;
        return p.x >= min_x && p.x <= max_x;
    }
    return false;
}

static bool is_inside_polygon(Point p, Point* polygon, int n) {
    int intersections = 0;

    for (int i = 0; i < n; i++) {
        Point p1 = polygon[i];
        Point p2 = polygon[(i + 1) % n];

        if ((p1.y > p.y) != (p2.y > p.y)) {
            double intersect_x = (double)(p2.x - p1.x) * (p.y - p1.y) / (p2.y - p1.y) + p1.x;
            if (p.x < intersect_x) {
                intersections++;
            }
        }
    }

    return (intersections % 2) == 1;
}

static bool is_inside_or_on_boundary(Point p, Point* polygon, int n) {
    for (int i = 0; i < n; i++) {
        if (is_point_on_segment(p, polygon[i], polygon[(i + 1) % n])) {
            return true;
        }
    }
    return is_inside_polygon(p, polygon, n);
}

int64_t day09_part1(const char* input) {
    char* data = strdup(input);
    if (!data) return 0;

    Point points[10000];
    int n = 0;

    char* line = strtok(data, "\r\n");
    while (line && n < 10000) {
        int x, y;
        if (sscanf(line, "%d,%d", &x, &y) == 2) {
            points[n].x = x;
            points[n].y = y;
            n++;
        }
        line = strtok(NULL, "\r\n");
    }

    free(data);

    if (n < 2) return 0;

    int64_t max_area = 0;

    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            int width = abs(points[j].x - points[i].x) + 1;
            int height = abs(points[j].y - points[i].y) + 1;
            int64_t area = (int64_t)width * height;

            if (area > max_area) {
                max_area = area;
            }
        }
    }

    return max_area;
}

int64_t day09_part2(const char* input) {
    char* data = strdup(input);
    if (!data) return 0;

    Point points[10000];
    int n = 0;

    char* line = strtok(data, "\r\n");
    while (line && n < 10000) {
        int x, y;
        if (sscanf(line, "%d,%d", &x, &y) == 2) {
            points[n].x = x;
            points[n].y = y;
            n++;
        }
        line = strtok(NULL, "\r\n");
    }

    free(data);

    if (n < 2) return 0;

    int64_t max_area = 0;

    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            int min_x = (points[i].x < points[j].x) ? points[i].x : points[j].x;
            int max_x = (points[i].x > points[j].x) ? points[i].x : points[j].x;
            int min_y = (points[i].y < points[j].y) ? points[i].y : points[j].y;
            int max_y = (points[i].y > points[j].y) ? points[i].y : points[j].y;

            // Check corners
            Point corners[4] = {
                {min_x, min_y},
                {min_x, max_y},
                {max_x, min_y},
                {max_x, max_y}
            };

            bool all_inside = true;
            for (int c = 0; c < 4; c++) {
                if (!is_inside_or_on_boundary(corners[c], points, n)) {
                    all_inside = false;
                    break;
                }
            }
            if (!all_inside) continue;

            // Check for interior tiles
            bool has_interior = false;
            for (int k = 0; k < n; k++) {
                if (points[k].x > min_x && points[k].x < max_x &&
                    points[k].y > min_y && points[k].y < max_y) {
                    has_interior = true;
                    break;
                }
            }
            if (has_interior) continue;

            // Check for boundary crossings
            bool has_crossing = false;
            Point rect_corners[4] = {
                {min_x, min_y},
                {max_x, min_y},
                {max_x, max_y},
                {min_x, max_y}
            };

            for (int k = 0; k < n && !has_crossing; k++) {
                Point p1 = points[k];
                Point p2 = points[(k + 1) % n];

                for (int r = 0; r < 4; r++) {
                    if (segments_properly_intersect(p1, p2, rect_corners[r], rect_corners[(r + 1) % 4])) {
                        has_crossing = true;
                        break;
                    }
                }
            }

            if (!has_crossing) {
                int64_t width = max_x - min_x + 1;
                int64_t height = max_y - min_y + 1;
                int64_t area = width * height;
                if (area > max_area) {
                    max_area = area;
                }
            }
        }
    }

    return max_area;
}
